EIPS/eip-5050.md

```diff
@@ -0,0 +1,323 @@
+- State contracts can be bridges to state contracts on another chain, allowing for L1-verification, L2-storage usage pattern (validate action with layer-1 assets, save on l2 where storage is cheaper).
+
+#### Example
+
+State Contract `FightGame` defines a fighting game environment. Token holders call `FightGame.register(contract, tokenId)` to randomly initialize their stats (strength/hp/etc.). An account which holds a registered token A of contract `Fighters`, calls `Fighters.sendAction(AttackAction)`, specifying token A from `Fighters` as the sender, token B from `Pacifists` contract as the receiver, and `FightGame` as the state contract.
+
+The action is passed to token B, which may handle the action in whatever way it wants before passing the action to the `FightGame` state contract. The state contract can verify the stored action hash with the `Fighters` contract to validate the action is authentic before updating the stats if the tokens, dealing damage to token B.
+
+Tokens A and B may update their metadata based on stats in the `FightGame` state contract, or based on their own stored data updated in response to sending/receiving actions.
+
+### Extensions
+
+#### Interactive
+
+Some contracts may have custom user interfaces that facilitate interaction.
+
+```solidity
+pragma solidity ^0.8.0;
+
+interface IERC5050Interactive {
+    function interfaceURI(bytes4 _action) external view returns (string);
+}
+```
+
+#### Action Proxies
+
+Action proxies can be used to support backwards compatibility with non-upgradeable contracts, and potentially for cross-chain action bridging.
+
+They can be implemented using a modified version of [EIP-1820](./eip-1820.md#erc-1820-registry-smart-contract) that allows [EIP-173](./eip-173.md) contract owners to call `setManager()`.
+
+#### Controllable
+
+Users of this standard may want to allow trusted contracts to control the action process to provide security guarantees, and support action bridging. Controllers step through the action chain, calling each contract individually in sequence.
+
+Contracts that support Controllers SHOULD ignore require/revert statements related to action verification, and MUST NOT pass the action to the next contract in the chain.
+
+```solidity
+pragma solidity ^0.8.0;
+
+interface IControllable {
+    event ControllerApproval(
+        address indexed _controller,
+        bytes4 indexed _action
+    );
+    
+    event ControllerApprovalForAll(
+        address indexed _controller,
+        bool _approved
+    );
+    
+    function approveController(address _controller, bytes4 _action)
+        external;
+
+    function setControllerApprovalForAll(address _controller, bool _approved)
+        external;
+
+    function isApprovedController(address _controller, bytes4 _action)
+        external
+        view
+        returns (bool);
```

Comment:
These should be described, e.g. using NatSpec.

Reply:
comments added in latest commit 👍